src/protocol/vm/state.rs

```diff
@@ -279,25 +286,34 @@ impl VMPoolState<PreCachedDB> {
 
         let sim_result = engine
             .simulate(&sim_params)
-            .map_err(TychoSimulationError::SimulationFailure)?;
+            .map_err(|err| TychoSimulationError::from(VMError::SimulationFailure(err)))?;
 
-        let address = decode(&[ParamType::Address], &sim_result.result)?
+        let address = decode(&[ParamType::Address], &sim_result.result)
+            .map_err(|_| {
+                TychoSimulationError::from(VMError::DecodingError("Failed to decode ABI".into()))
+            })?
             .into_iter()
             .next()
             .ok_or_else(|| {
-                TychoSimulationError::DecodingError("Expected an address in the result".into())
+                TychoSimulationError::from(VMError::DecodingError(
+                    "Expected an address in the result".into(),
```

Comment:
Same with these error messages with the expect wording. To an outside user this will make no sense... the engine raises an error that some sort of internal result had no address in it? It's not clear to me how to fix that or what actually went wrong.